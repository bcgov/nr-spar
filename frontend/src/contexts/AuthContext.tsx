import { KeycloakLoginOptions } from 'keycloak-js';
import React, {
  createContext,
  useContext,
  useMemo,
  useState
} from 'react';
import KeycloakService from '../service/KeycloakService';
import KeycloakUser from '../types/KeycloakUser';

interface AuthContextData {
  signed: boolean;
  user: KeycloakUser | {};
  startKeycloak(): Promise<boolean>;
  login(options?: KeycloakLoginOptions): Promise<void>;
  logout(): Promise<void>;
  createLoginUrl(options?: KeycloakLoginOptions): string;
  provider: string;
  token: string | undefined,
}

const AuthContext = createContext<AuthContextData>({} as AuthContextData);

const REFRESH_TIMER = 2 * 60 * 1000;

interface Props {
  children: React.ReactNode;
}

const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }: Props) => {
  const [signed, setSigned] = useState<boolean>(false);
  const [user, setUser] = useState<KeycloakUser | {}>({});

  /**
   * Starts Keycloak instance.
   */
  async function startKeycloak() {
    try {
      const userIsLoggedIn = await KeycloakService.initKeycloak();
      setSigned(userIsLoggedIn);
      const kcUser = KeycloakService.getUser();
      setUser(kcUser);
      return userIsLoggedIn;
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log('Keycloak init error:', e);
    }
    return false;
  }

  /**
   * Logout the user
   */
  async function logout() {
    try {
      localStorage.clear();
      await KeycloakService.logout();
      setUser({});
      setSigned(false);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('Keycloak logout error:', e);
    }
  }

  /**
   * Refresh the token
   */
  setInterval(() => {
    KeycloakService.updateToken(30)
      .catch(async (err) => {
        // eslint-disable-next-line no-console
        console.error('Keycloack service update error: ', err);
        await logout();
      });
  }, REFRESH_TIMER);

  const { createLoginUrl, login } = KeycloakService;
  const provider = KeycloakService.authMethod();
  const token = KeycloakService.getToken();

  // memoize
  const contextValue = useMemo(() => ({
    signed,
    user,
    startKeycloak,
    login,
    logout,
    createLoginUrl,
    provider,
    token
  }), [signed, user, startKeycloak, login, logout, createLoginUrl, provider, token]);

  return (
    <AuthContext.Provider value={contextValue}>
      {children}
    </AuthContext.Provider>
  );
};

/**
 * Create useAuth hook.
 *
 * @returns {AuthContext} context.
 */
function useAuth(): AuthContextData {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error('useAuth must be used within an AuthProvider');
  }

  return context;
}

export { AuthProvider, useAuth };
