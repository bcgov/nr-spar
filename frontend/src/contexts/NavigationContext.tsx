import React, { createContext, useContext, useMemo } from 'react';

interface NavigateContextType {
  redirectTo403: Function;
}

const NavigationContext = createContext<NavigateContextType | undefined>(undefined);

// This a navigation provider, where a a context will be used to access the useNavigate
// without necessarily being inside a router context
export const NavigateProvider: React.FC<{
  children: React.ReactNode,
  onRedirect: () => void
}> = ({ children, onRedirect }) => {
  const value = useMemo(() => ({ redirectTo403: onRedirect }), [onRedirect]);

  return (
    <NavigationContext.Provider value={value}>
      {children}
    </NavigationContext.Provider>
  );
};

export const useNavigateContext = () => {
  const context = useContext(NavigationContext);

  if (!context) {
    throw new Error('useNavigateContext must be used within a NavigateProvider');
  }

  return context;
};
