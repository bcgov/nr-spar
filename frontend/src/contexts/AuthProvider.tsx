/* eslint-disable no-console */
import React, {
  useMemo,
  useState
} from 'react';
import type { CognitoUserSession } from 'amazon-cognito-identity-js';
import { Auth } from 'aws-amplify';
import axios from 'axios';
import { env } from '../env';
import FamUser from '../types/FamUser';
import LoginProviders from '../types/LoginProviders';
import AuthContext from './AuthContext';
import { SPAR_REDIRECT_PATH } from '../shared-constants/shared-constants';

interface Props {
  children: React.ReactNode;
}

const FAM_LOGIN_USER = 'famLoginUser';

const findFindAndLastName = (displayName: string, provider: string): Array<string> => {
  let splitString = ', ';
  if (!displayName.includes(splitString)) {
    splitString = ' ';
  }
  const parts = displayName.split(splitString);
  let firstName = parts[0];
  let lastName = parts[1];

  if (parts.length > 2) {
    lastName = `${lastName} ${parts[2]}`;
  }

  if (provider === LoginProviders.IDIR) {
    firstName = parts[1].split(' ')[0].trim();
    lastName = parts[0].includes(' ') ? parts[0].split(' ')[1] : parts[0];
  }

  return [firstName, lastName];
};

/**
 * Parses a CognitoUserSession into a JS object. For a deeper understanding
 * you can take a look on the attribute mapping reference at:
 * https://github.com/bcgov/nr-forests-access-management/wiki/OIDC-Attribute-Mapping
 *
 * @param authToken The CognitoUserSession to be parsed.
 * @returns The FamUser object
 */
const parseToken = (authToken: CognitoUserSession): FamUser => {
  const decodedIdToken = authToken.getIdToken().decodePayload();
  const decodedAccessToken = authToken.getAccessToken().decodePayload();

  // Extract the first name and last name from the displayName and remove unwanted part
  const displayName = decodedIdToken['custom:idp_display_name'] as string;
  const provider = decodedIdToken['custom:idp_name'] as string;
  const [firstName, lastName] = findFindAndLastName(displayName, provider.toUpperCase());
  const famUser: FamUser = {
    displayName: decodedIdToken['custom:idp_display_name'], // E.g: 'de Campos, Ricardo WLRS:EX'
    email: decodedIdToken.email,
    lastName,
    firstName,
    providerUsername: decodedIdToken['custom:idp_username'], // E.g: RDECAMPO
    name: `${firstName} ${lastName}`,
    roles: decodedAccessToken['cognito:groups'],
    provider: decodedIdToken['custom:idp_name'].toLocaleUpperCase(),
    jwtToken: authToken.getIdToken().getJwtToken(),
    refreshToken: authToken.getRefreshToken().getToken(),
    axiosRequestUser: decodedAccessToken.username, // E.g: 'dev-idir_***@idir'
    tokenExpiration: authToken.getIdToken().getExpiration()
  };

  return famUser;
};

const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }: Props) => {
  const [signed, setSigned] = useState<boolean>(false);
  const [user, setUser] = useState<FamUser | null>(null);
  const [provider, setProvider] = useState<string>('');
  const [token, setToken] = useState<string>('');
  const [intervalInstance, setIntervalInstance] = useState<NodeJS.Timeout | null>(null);

  const fetchFamCurrentSession = async (pathname: string): Promise<FamUser | null> => {
    try {
      const currentSession: CognitoUserSession = await Auth.currentSession();
      const famUser = parseToken(currentSession);
      setSigned(true);
      return famUser;
    } catch (e) {
      console.warn(e);
      localStorage.setItem(SPAR_REDIRECT_PATH, pathname);
      setSigned(false);
    }
    return null;
  };

  const updateUserFamSession = (famUser: FamUser) => {
    localStorage.setItem(FAM_LOGIN_USER, JSON.stringify(famUser));

    // axios token
    axios.defaults.headers.common = {
      // Authorization: `Bearer ${token}`
      'Temporary-User-Identification': famUser.axiosRequestUser
    };
  };

  const isCurrentAuthUser = async (pathname: string) => {
    const famUser = await fetchFamCurrentSession(pathname);
    if (famUser) {
      updateUserFamSession(famUser);
      setUser(famUser);
      setProvider(famUser.provider);
      setToken(famUser.axiosRequestUser);
    }
  };

  const signIn = (signInProvider: LoginProviders): void => {
    const appEnv = env.VITE_ZONE ?? 'DEV';

    Auth.federatedSignIn({
      customProvider: `${(appEnv).toLocaleUpperCase()}-${signInProvider.toString()}`
    });
  };

  const signOut = (): void => {
    Auth.signOut()
      .then(() => {
        localStorage.removeItem(FAM_LOGIN_USER);
        setSigned(false);
        setUser(null);
        setProvider('');
        setToken('');
        if (intervalInstance) {
          console.log('stopping refresh token');
          clearInterval(intervalInstance);
          setIntervalInstance(null);
        }
      }).catch((err) => console.warn(err));
  };

  const refreshTokenPvt = async () => {
    const famUser = await fetchFamCurrentSession('/');
    if (famUser) {
      updateUserFamSession(famUser);
    }
  };

  // 2 minutes
  const REFRESH_TIMER = 2 * 60 * 1000;

  if (intervalInstance == null && signed) {
    const instance = setInterval(() => {
      refreshTokenPvt()
        .then(() => {
          console.log('User session successfully refreshed!');
        })
        .catch((e) => console.error(e));
    }, REFRESH_TIMER);

    setIntervalInstance(instance);
  }

  // memoize
  const contextValue = useMemo(() => ({
    signed,
    user,
    isCurrentAuthUser,
    signIn,
    signOut,
    provider,
    token
  }), [signed, user, isCurrentAuthUser, signIn, signOut, provider, token]);

  return (
    <AuthContext.Provider value={contextValue}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
