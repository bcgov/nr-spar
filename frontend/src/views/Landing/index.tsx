import React, { useEffect } from 'react';

import {
  Button,
  Grid,
  Column
} from '@carbon/react';
import { Login } from '@carbon/icons-react';
import { KeycloakLoginOptions } from 'keycloak-js';
import { useNavigate } from 'react-router-dom';

import { useAuth } from '../../contexts/AuthContext';
import BCGovLogo from '../../components/BCGovLogo';
import Seeding from '../../assets/img/seeding.png';

import LoginProviders from '../../types/LoginProviders';

import './styles.scss';

import getUrlQueryParam from '../../utils/UrlUtils';

const Landing = () => {
  const { startKeycloak, login, signed } = useAuth();
  const navigate = useNavigate();
  const homePage = '/dashboard';

  const handleLogin = (provider: LoginProviders) => {
    if (signed) {
      navigate(getUrlQueryParam(window.location, 'page') || homePage);
      return;
    }

    const idpHint = provider;
    const loginOptions: KeycloakLoginOptions = {
      redirectUri: `${window.location.origin}${homePage}`,
      idpHint
    };

    login(loginOptions);
  };

  useEffect(() => {
    if (signed) {
      navigate(getUrlQueryParam(window.location, 'page') || homePage);
    } else {
      startKeycloak();
    }
  }, [signed]);

  return (
    <Grid fullWidth className="landing-grid">
      {/* First - Column */}
      <Column sm={4} md={5} lg={10}>
        {/* Logo */}
        <BCGovLogo />

        {/* Welcome - Title and Subtitle */}
        <h1 data-testid="landing-title" className="landing-title">Welcome to SPAR</h1>
        <h2 data-testid="landing-subtitle" className="landing-subtitle">
          Seed Planning and Registry Application
        </h2>

        {/* Description */}
        <p data-testid="landing-desc" className="landing-desc">
          Register and storage your seed and meet your annual
          reforestation needs using
          <span className="spar-span"> SPAR</span>
        </p>

        {/* Login buttons */}
        <Button
          onClick={() => { handleLogin(LoginProviders.IDIR); }}
          size="md"
          renderIcon={Login}
          data-testid="landing-button__idir"
          className="btn-landing"
        >
          Login with IDIR
        </Button>

        <Button
          kind="tertiary"
          onClick={() => { handleLogin(LoginProviders.BCEID_BUSINESS); }}
          size="md"
          renderIcon={Login}
          data-testid="landing-button__bceid"
          className="btn-landing"
        >
          Login with Business BCeID
        </Button>
      </Column>

      {/* Second - Column */}
      <Column className="seeding-img-column" sm={4} md={3} lg={6}>
        <img
          src={Seeding}
          alt="Small green seedling on the dirt and watered"
          className="seeding-img"
        />
      </Column>
    </Grid>
  );
};

export default Landing;
