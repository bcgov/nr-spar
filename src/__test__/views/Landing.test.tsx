/* eslint-disable no-undef */
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Landing from '../../views/Landing';
import { AuthProvider } from '../../contexts/AuthContext';

describe('Landing component test', () => {
  it('should have the correct title', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <AuthProvider>
          <Landing />
        </AuthProvider>
      </BrowserRouter>
    );

    const content = {
      title: 'Welcome to SPAR',
      subtitle: 'Seed Planning and Registry Application',
      description: 'Register and storage your seed and meet '
        + 'your annual reforestation needs using SPAR',
      idir_btn: 'Login with IDIR',
      bceid_btn: 'Login with Business BCeID'
    };

    expect(getByTestId('landing-title').textContent).toBe(content.title);
    expect(getByTestId('landing-subtitle').textContent).toBe(content.subtitle);
    expect(getByTestId('landing-desc').textContent).toBe(content.description);
    expect(getByTestId('landing-button__idir').textContent).toBe(content.idir_btn);
    expect(getByTestId('landing-button__bceid').textContent).toBe(content.bceid_btn);
  });

  it('should match the snapshot', () => {
    const landing = renderer
      .create(
        <BrowserRouter>
          <AuthProvider>
            <Landing />
          </AuthProvider>
        </BrowserRouter>
      ).toJSON();

    expect(landing).toMatchSnapshot();
  });
});
