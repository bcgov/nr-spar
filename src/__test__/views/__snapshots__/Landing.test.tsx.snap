// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing component test should match the snapshot 1`] = `
<div
  className="landing-grid cds--css-grid cds--css-grid--full-width"
>
  <div
    className="cds--sm:col-span-4 cds--md:col-span-5 cds--lg:col-span-10 cds--css-grid-column"
  >
    <img
      alt="BCGov Logo"
      className="logo"
      src=""
      width={160}
    />
    <h1
      className="landing-title"
      data-testid="landing-title"
    >
      Welcome to SPAR
    </h1>
    <h2
      className="landing-subtitle"
      data-testid="landing-subtitle"
    >
      Seed Planning and Registry Application
    </h2>
    <p
      className="landing-desc"
      data-testid="landing-desc"
    >
      Register and storage your seed and meet your annual reforestation needs using
      <span
        className="spar-span"
      >
         SPAR
      </span>
    </p>
    <button
      aria-describedby={null}
      aria-pressed={null}
      className="btn-landing cds--btn cds--btn--md cds--btn--primary"
      data-testid="landing-button__idir"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="button"
    >
      Login with IDIR
      <svg
        aria-hidden={true}
        className="cds--btn__icon"
        fill="currentColor"
        focusable="false"
        height={16}
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 32 32"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26,30H14a2,2,0,0,1-2-2V25h2v3H26V4H14V7H12V4a2,2,0,0,1,2-2H26a2,2,0,0,1,2,2V28A2,2,0,0,1,26,30Z"
        />
        <path
          d="M14.59 20.59L18.17 17 4 17 4 15 18.17 15 14.59 11.41 16 10 22 16 16 22 14.59 20.59z"
        />
      </svg>
    </button>
    <button
      aria-describedby={null}
      aria-pressed={null}
      className="btn-landing cds--btn cds--btn--md cds--btn--tertiary"
      data-testid="landing-button__bceid"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="button"
    >
      Login with Business BCeID
      <svg
        aria-hidden={true}
        className="cds--btn__icon"
        fill="currentColor"
        focusable="false"
        height={16}
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 32 32"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26,30H14a2,2,0,0,1-2-2V25h2v3H26V4H14V7H12V4a2,2,0,0,1,2-2H26a2,2,0,0,1,2,2V28A2,2,0,0,1,26,30Z"
        />
        <path
          d="M14.59 20.59L18.17 17 4 17 4 15 18.17 15 14.59 11.41 16 10 22 16 16 22 14.59 20.59z"
        />
      </svg>
    </button>
  </div>
  <div
    className="seeding-img-column cds--sm:col-span-4 cds--md:col-span-3 cds--lg:col-span-6 cds--css-grid-column"
  >
    <img
      alt="Small green seedling on the dirt and watered"
      className="seeding-img"
      src=""
    />
  </div>
</div>
`;
